File: String/08_StringtoInteger.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


class Solution(object):
    MAX_INT = 2**31 - 1
    MIN_INT = - 2**31

    def myAtoi(self, strs):
        """ We need to handle four cases:

        1. discards all leading whitespaces
        2. sign of the number
        3. overflow
        4. invalid input
        """
        strs = strs.strip()
        if not strs:
            return 0

        sign, i = 1, 0
        if strs[i] == '+':
            i += 1
        elif strs[i] == '-':
            i += 1
            sign = -1

        num = 0
        while i < len(strs):
            if strs[i] < '0' or strs[i] > '9':
                break
            if num > self.MAX_INT or (num * 10 + int(strs[i]) > self.MAX_INT):
                return self.MAX_INT if sign == 1 else self.MIN_INT
            else:
                num = num * 10 + int(strs[i])
            i += 1

        return num * sign

"""
""
"  12a"
"  a12"
"  +12"
"  +-12"
"2147483648"
"-2147483648"
"""

File: String/08_StringtoInteger.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-08-21 09:10:58
 */

class Solution {
public:
    int myAtoi(string str){
        /*
        We need to handle four cases:

        1. discards all leading whitespaces
        2. sign of the number
        3. overflow
        4. invalid input
        */
        const int len = str.size();

        int i=0;
        // Skip starting whitespace as many as possible.
        while (str[i]==' ' && i<len) i++;

        int sign = 1;
        // Read plus or minus sign.
        if(str[i] == '+') i++;
        else if(str[i] == '-'){
            i++;
            sign = -1;
        }

        int num = 0;
        for(; i<len; i++){
            if(str[i] < '0' || str[i] > '9'){
                break;
            }
            if(num > INT_MAX/10 || (num == INT_MAX / 10 && (str[i]-'0') > INT_MAX % 10)){
                return sign == -1 ? INT_MIN : INT_MAX;
            }
            num = num * 10 + str[i] - '0';
        }

        return num * sign;
    }
};

/*
""
"  12a"
"  a12"
"  +12"
"  +-12"
"2147483648"
"-2147483648"
*/
